lib/matplotlib/__init__.py

```diff
@@ -761,30 +818,147 @@ def __getitem__(self, key):
+        try:
+            if depth == 1:
+                if key in self.single_key_set:
+                    del self._namespace_maps["default"][key]
+                else:
+                    raise KeyError
+            elif depth == 2:
+                del self._namespace_maps[keys[0]][keys[1]]
+        except KeyError as err:
+            raise KeyError(
+                f"{key} is not a valid rc parameter (see rcParams.keys() for "
+                f"a list of valid parameters)") from err
+
+    def __contains__(self, key):
+        keys, depth = self._split_key(key)
+        if depth == 1:
+            if key in self.single_key_set:
+                return key in self._namespace_maps["default"]
+            else:
+                return False
+        elif depth == 2:
+            return any(key in mapping for mapping in self._namespace_maps)
+
+    def __iter__(self):
+        """Yield from sorted list of keys"""
+        keys = (
+            ".".join((space, key)) if space != 'default' else key
+            for space, mapping in self._namespace_maps.items()
+            for key in mapping.keys()
+        )
+        with _api.suppress_matplotlib_deprecation_warning():
+            yield from sorted(keys)
+
+    def __len__(self):
+        return sum(len(mapping) for mapping in self._namespace_maps)
+
     def __repr__(self):
         class_name = self.__class__.__name__
         indent = len(class_name) + 1
         with _api.suppress_matplotlib_deprecation_warning():
-            repr_split = pprint.pformat(dict(self), indent=1,
+            repr_split = pprint.pformat(dict(self.items()), indent=1,
                                         width=80 - indent).split('\n')
         repr_indented = ('\n' + ' ' * indent).join(repr_split)
         return f'{class_name}({repr_indented})'
 
     def __str__(self):
         return '\n'.join(map('{0[0]}: {0[1]}'.format, sorted(self.items())))
 
-    def __iter__(self):
-        """Yield sorted list of keys."""
-        with _api.suppress_matplotlib_deprecation_warning():
-            yield from sorted(dict.__iter__(self))
+    def pop(self, key):
+        keys, depth = self._split_key(key)
+        if depth == 1:
+            if key in self.single_key_set:
+                return self._namespace_mapping["default"][key]
+            else:
+                raise KeyError(
+                    f"{key} is not a valid rc parameter (see rcParams.keys() for "
+                    f"a list of valid parameters)")
+        elif depth == 2:
+            return self._namespace_mapping[keys[0]].pop(keys[1])
 
-    def __len__(self):
-        return dict.__len__(self)
+    def popitem(self):
+        raise NotImplementedError(
+            "popitem is not implemented for RcParams.")
+
+    def clear(self):
+        for namespace in self._namespace_maps:
+            self._namespace_maps[namespace].clear()
+
+    def setdefault(self, key, default=None):
+        self[key] = default
+        return default
```

Comment:
This is not what `dict.setdefault` does. (set *key* to *default* if *key* is not in the dict).

In fact, our current implementation does not allow adding custom keys. We raise on setdefault if *key* is not a known key. And we return the value if key is known; as a result `rcParams.setdefault(key, ...)` is equivalent to `rcParams[key]`. and `setdefault` is not really needed. - Separate question is whether we can throw it out immediately or have to deprecate it.

Reply:
Yes, so sorry. Not sure how I missed that. I'll change that.